go/parquet/internal/encoding/byte_stream_split.go

```diff
@@ -0,0 +1,376 @@
+func encodeByteStreamSplit(data []byte, in []byte, numElements, width int) {
+	for stream := 0; stream < width; stream++ {
+		for element := 0; element < numElements; element++ {
+			encLoc := numElements*stream + element
+			decLoc := width*element + stream
+			data[encLoc] = in[decLoc]
+		}
+	}
+}
+
+// encodeByteStreamSplitWidth4 implements encodeByteStreamSplit optimized for types stored using 4 bytes
+func encodeByteStreamSplitWidth4(data []byte, in []byte, numElements int) {
+	width := 4
+	for element := 0; element < numElements; element++ {
+		decLoc := width * element
+		data[element] = in[decLoc]
+		data[numElements+element] = in[decLoc+1]
+		data[numElements*2+element] = in[decLoc+2]
+		data[numElements*3+element] = in[decLoc+3]
+	}
+}
+
+// encodeByteStreamSplitWidth8 implements encodeByteStreamSplit optimized for types stored using 8 bytes
+func encodeByteStreamSplitWidth8(data []byte, in []byte, numElements int) {
+	width := 8
+	for element := 0; element < numElements; element++ {
+		decLoc := width * element
+		data[element] = in[decLoc]
+		data[numElements+element] = in[decLoc+1]
+		data[numElements*2+element] = in[decLoc+2]
+		data[numElements*3+element] = in[decLoc+3]
+		data[numElements*4+element] = in[decLoc+4]
+		data[numElements*5+element] = in[decLoc+5]
+		data[numElements*6+element] = in[decLoc+6]
+		data[numElements*7+element] = in[decLoc+7]
+	}
+}
+
+// putByteStreamSplitSpaced encodes data that has space for nulls using the BYTE_STREAM_SPLIT encoding, calling to the provided putFn to encode runs of non-null values.
+func putByteStreamSplitSpaced[T ByteStreamSplitType](in []T, validBits []byte, validBitsOffset int64, bitSetReader bitutils.SetBitRunReader, putFn func([]T)) {
+	if validBits == nil {
+		putFn(in)
+		return
+	}
+
+	if bitSetReader == nil {
+		bitSetReader = bitutils.NewSetBitRunReader(validBits, validBitsOffset, int64(len(in)))
+	} else {
+		bitSetReader.Reset(validBits, validBitsOffset, int64(len(in)))
+	}
+
+	for {
+		run := bitSetReader.NextRun()
+		if run.Length == 0 {
+			break
+		}
+		putFn(in[int(run.Pos):int(run.Pos+run.Length)])
+	}
+}
+
+// decodeByteStreamSplit is a generic implementation of the BYTE_STREAM_SPLIT decoder for fixed-width numeric types.
+func decodeByteStreamSplitNumeric[T NumericByteStreamSplitType](out []T, dec TypedDecoder, data []byte) (int, error) {
+	max := utils.Min(len(out), dec.ValuesLeft())
+	typeLen := dec.Type().ByteSize()
+	numBytesNeeded := max * typeLen
+	if numBytesNeeded > len(data) || numBytesNeeded > math.MaxInt32 {
+		return 0, xerrors.New("parquet: eof exception")
+	}
+
+	outBytes := arrow.GetBytes(out)
+	switch typeLen {
+	case 4:
+		decodeByteStreamSplitWidth4(data, outBytes, max)
+	case 8:
+		decodeByteStreamSplitWidth8(data, outBytes, max)
+	default:
+		decodeByteStreamSplit(data, outBytes, max, typeLen)
+	}
+
+	dec.SetData(dec.ValuesLeft()-max, data[numBytesNeeded:])
```

Comment:
Yea i think you're gonna run into an issue here in general with subsequent calls. the "number of elements" should always be the *total* number of elements in the data page from what I understand, so if someone makes multiple calls to decode here for grabbing small sections of data, or tries to skip, you're going to get incorrect values I think with the current implementation